File: 160.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB:
            return None
        a,b = headA,headB
        lena,lenb = 0,0
        while a:
            lena += 1
            a = a.next
        while b:
            lenb += 1
            b = b.next
        a,b = headA,headB
        if lena > lenb:
            for i in range(lena-lenb):
                a = a.next
        elif lenb > lena:
            for i in range(lenb-lena):
                b = b.next
        while a != b:
            a = a.next 
            b = b.next 
        return a

File: 160.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        ListNode n1 = headA;
        ListNode n2 = headB;
        while (n1 != n2) {
            if (n1.next == null && n2.next == null) return null;
            if (n1.next == null) n1 = headB;
            else n1 = n1.next;
            if (n2.next == null) n2 = headA;
            else n2 = n2.next;
        }
        return n1;
    }
}